File: algorithms/medium/count_number_of_maximum_bitwise_or_subsets.ex
defmodule Solution do
  use Bitwise

  @spec count_max_or_subsets(nums :: [integer]) :: integer
  def count_max_or_subsets(nums) do
    count_max_or_subsets(nums, 0, -1, 0) |> elem(1)
  end

  def count_max_or_subsets(nums, cur, max, ans) do
    {max, ans} =
      cond do
        cur > max -> {cur, 1}
        cur == max -> {max, ans + 1}
        true -> {max, ans}
      end

    if nums == [] do
      {max, ans}
    else
      for i <- 0..(Enum.count(nums) - 1), reduce: {max, ans} do
        {max, ans} ->
          [x | xs] = Enum.drop(nums, i)
          count_max_or_subsets(xs, bor(cur, x), max, ans)
      end
    end
  end
end

File: algorithms/medium/count_number_of_maximum_bitwise_or_subsets.rs
fn dfs(nums: &[i32], cur: i32, max: &mut i32, ans: &mut i32) {
    if cur > *max {
        *max = cur;
        *ans = 1;
    } else if cur == *max {
        *ans += 1;
    }

    for (i, &n) in nums.iter().enumerate() {
        dfs(&nums[i + 1..], cur | n, max, ans)
    }
}

impl Solution {
    pub fn count_max_or_subsets(nums: Vec<i32>) -> i32 {
        let mut max = -1;
        let mut ans = 0;
        dfs(&nums, 0, &mut max, &mut ans);
        ans
    }
}
